src/daemon/scheduler.rs

```diff
@@ -1,145 +1,101 @@
-    /// entitlements have changed *and* for the shrinking of issued certificates, if
-    /// they are not renewed within the configured grace period.
-    #[allow(dead_code)] // just need to keep this in scope
-    cas_refresh: ScheduleHandle,
-
-    /// Responsible for refreshing announcement information
-    #[allow(dead_code)] // just need to keep this in scope
-    announcements_refresh: ScheduleHandle,
-
+    tasks: Arc<TaskQueue>,
+    ca_manager: Arc<CaManager>,
+    bgp_analyser: Arc<BgpAnalyser>,
     #[cfg(feature = "multi-user")]
-    /// Responsible for purging expired cached login tokens
-    #[allow(dead_code)] // just need to keep this in scope
-    login_cache_sweeper_sh: ScheduleHandle,
+    // Responsible for purging expired cached login tokens
+    login_session_cache: Arc<LoginSessionCache>,
+    config: Arc<Config>,
+    system_actor: Actor,
+    started: Timestamp,
 }
 
 impl Scheduler {
     pub fn build(
-        event_queue: Arc<MessageQueue>,
+        tasks: Arc<TaskQueue>,
         ca_manager: Arc<CaManager>,
         bgp_analyser: Arc<BgpAnalyser>,
         #[cfg(feature = "multi-user")] login_session_cache: Arc<LoginSessionCache>,
-        config: &Config,
-        actor: &Actor,
+        config: Arc<Config>,
+        system_actor: Actor,
     ) -> Self {
-        let cas_event_triggers = make_cas_event_triggers(event_queue.clone(), ca_manager.clone(), actor.clone());
-
-        let cas_republish = make_cas_republish(ca_manager.clone(), event_queue);
-        let cas_objects_renew = make_cas_objects_renew(ca_manager.clone(), actor.clone());
-        let cas_refresh = make_cas_refresh(ca_manager, config.ca_refresh_seconds, actor.clone());
-
-        let announcements_refresh = make_announcements_refresh(bgp_analyser);
-
-        #[cfg(feature = "multi-user")]
-        let login_cache_sweeper_sh = make_login_cache_sweeper_sh(login_session_cache);
-
         Scheduler {
-            cas_event_triggers,
-            cas_republish,
-            cas_objects_renew,
-            cas_refresh,
-            announcements_refresh,
+            tasks,
+            ca_manager,
+            bgp_analyser,
             #[cfg(feature = "multi-user")]
-            login_cache_sweeper_sh,
+            login_session_cache,
+            config,
+            system_actor,
+            started: Timestamp::now(),
         }
     }
-}
 
-#[allow(clippy::cognitive_complexity)]
-fn make_cas_event_triggers(event_queue: Arc<MessageQueue>, ca_manager: Arc<CaManager>, actor: Actor) -> ScheduleHandle {
-    let started = Timestamp::now();
+    /// Run the schedular in the background. It will sweep the message queue for tasks
+    /// and re-schedule new tasks as needed.
+    pub async fn run(&self) -> KrillResult<()> {
+        loop {
+            while let Some(evt) = self.tasks.pop(now()) {
+                match evt {
+                    Task::ServerStarted => self.queue_start_tasks().await?, // return error and stop server on failure
```

Comment:
I've tried to compare the old scheduler setup to the new priority task setup and came up with this summary:

Task::CheckSuspendChildren: NEW: Wasn't periodic, is now?
Task::RefreshAnnouncementsInfo: CHANGED: Was cron every 5 seconds, but is now a recurring task on the queue every 10 minutes?
Task::RenewObjectsIfNeeded: UNCHANGED
Task::RepublishIfNeeded: CHANGED: Was cron every 10 minutes, but is now a recurring task on the queue every 5 minutes?
QueueTask::RescheduleSyncParent: WAS UNUSED SO REMOVED?
QueueTask::RescheduleSyncRepo: WAS UNUSED SO REMOVED?
QueueTask::ResourceClassRemoved: UNCHANGED
QueueTask::ServerStarted: UNCHANGED
Task::SweepLoginCache: UNCHANGED
QueueTask::SyncParent: UNCHANGED
QueueTask::SyncRepo: UNCHANGED
QueueTask::UnexpectedKey: UNCHANGED

Reply:
Let me go through them one by one:

```Task::CheckSuspendChildren: NEW: Wasn't periodic, is now?```

It was periodic, but done as a side-effect. Now it has its own.

```Task::RefreshAnnouncementsInfo: CHANGED: Was cron every 5 seconds, but is now a recurring task on the queue every 10 minutes?```

This still uses another internal metric to avoid fetching the file every time. It will only do a refetch after 1 hour (and use if-modified-since). So checking every 5 seconds was excessive. I think every 10 minutes is fine given that these files are updated every 8 hours or so.

```Task::RenewObjectsIfNeeded: UNCHANGED```

ack

```Task::RepublishIfNeeded: CHANGED: Was cron every 10 minutes, but is now a recurring task on the queue every 5 minutes?```

It could also be 10 minutes, but every 5 minutes is fine as well. It's a cheap no-op in most case. That said, what I want to do in future is that instead of this a new actual mft/crl re-issuance is planned rather than checking every 5 minutes if work is needed.

```
QueueTask::RescheduleSyncParent: WAS UNUSED SO REMOVED?
QueueTask::RescheduleSyncRepo: WAS UNUSED SO REMOVED?
```

This was used to re-schedule those tasks with a 'due' time in case of failure. Now all tasks have a priority so this is not needed anymore. We just plan a normal sync parent/repo with the appropriate priority: try again in 15 minutes I think. 

```
Task::ResourceClassRemoved: UNCHANGED
Task::ServerStarted: UNCHANGED
Task::SweepLoginCache: UNCHANGED
Task::SyncParent: UNCHANGED
Task::SyncRepo: UNCHANGED
Task::UnexpectedKey: UNCHANGED
```

What may be interesting here is that `ServerStarted` is added when the server starts, and then it triggers that other tasks are planned. But because this is done later - this is asynchronous and does not delay the boot process. We could also just plan all tasks immediately without this. I don't really want to change this now, but I am not 100% sure what is best.